docs(tip-1065): update from meta-TIP to independent TIP (#4191)
tempoxyz/tempo: tips/tip-1065.md

Balance writes encode the new `balance` in the low 128 bits. Existing defined cache fields MAY be preserved, cleared, or updated according to their field-specific rules. Reserved bits MUST remain zero unless a future TIP assigns them. Future cache fields MAY use the reserved region only through a future TIP that defines their encoding, source of truth, invalidation behavior, and handling of invalid values. In practice this means that a packed state slot MAY have `balance == 0` while one of the cache fields is nonzero.

## Reward State Cache

`rewardFlag` is the first cache field defined in the user-state slot. It caches whether the account is opted into TIP-20 rewards so balance-changing paths can avoid unnecessary `SLOAD`s and mapping-key `keccak256` computations against reward-info storage. It can have the following values:

- `0`: `Uninitialized`.
- `1`: `OptedOut`.
- `2`: `OptedIn`.

> Any other encoded value is invalid and MUST be treated as a consensus error.

Reviewer: Does this mean if there was an invalid enum somehow we'd stop the chain? I think we'd just raise `PanicKind::EnumConversionError` in https://github.com/tempoxyz/tempo/pull/4161/

Author: initially i had it say it would be treated as `Uninitialized` but @dankrad made this correction.

In principle we only validate enums that are exposed via calldata. In this case the protocol impl should ensure no other values are possible.

`rewardRecipient(account)` remains the source of truth for reward delegation and opt-in status, and any initialized `rewardFlag` MUST always be aligned with it. As defined in TIP-20, `rewardRecipient != address(0)` means opted in, and `rewardRecipient == address(0)` means opted out.

> The reward-state cache may be used by balance-changing paths that need to know whether a holder is opted into rewards, including transfers, mints, fee refunds, reward claims, and any internal balance movement that updates `optedInSupply`. These paths MAY use an initialized `rewardFlag` instead of reading reward info from separate mapping slots. When `rewardFlag` is `Uninitialized` or invalid, the implementation MUST read `rewardRecipient(account)` and derive the effective reward state from it.

Reviewer: > Any other encoded value is invalid and MUST be treated as a consensus error.

> When `rewardFlag` is `Uninitialized` or invalid, the implementation MUST read `rewardRecipient(account)`

These two statements are at odds

> The reward-state cache may be used by balance-changing paths that need to know whether a holder is opted into rewards, including transfers, mints, fee refunds, reward claims, and any internal balance movement that updates `optedInSupply`. These paths MAY use an initialized `rewardFlag` instead of reading reward info from separate mapping slots. When `rewardFlag` is `Uninitialized` or invalid, the implementation MUST read `rewardRecipient(account)` and derive the effective reward state from it.

Author: ```suggestion
The reward-state cache may be used by balance-changing paths that need to know whether a holder is opted into rewards, including transfers, mints, fee refunds, reward claims, and any internal balance movement that updates `optedInSupply`. These paths MAY use an initialized `rewardFlag` instead of reading reward info from separate mapping slots. When `rewardFlag` is `Uninitialized`, the implementation MUST read `rewardRecipient(account)` and derive the effective reward state from it.
```

The first use case is TIP-20 reward accounting, which currently requires reading reward metadata from separate mapping slots during common balance-changing paths. Caching reward state in the existing balance slot reduces redundant storage reads and associated mapping hashing while keeping room in the same slot for future per-user caches.

## Assumptions

- TIP-20 total supply is capped at `uint128::MAX`, so each account balance fits in 128 bits.
- `rewardRecipient != address(0)` means opted in, and `rewardRecipient == address(0)` means opted out.

---

# Specification

At T6, each value in the TIP-20 `balance` mapping is interpreted as a packed `state` word instead of a raw `uint256` balance.

## User State Slot Layout

| Bits | Size | Field |
|---|---:|---|
| `0..127` | 128 | `balance` |
| `128..135` | 8 | `rewardFlag` |

> | `136..255` | 120 | reserved, MUST be zero |

Reviewer 2: ```suggestion
| `128..129` | 2 | `rewardFlag` |
| `130..255` | 126 | reserved, MUST be zero |
```
We only need 2 bits, let's preserve as much as possible as I think we can store a lot of other useful things here

`balance` is encoded as `uint128`. This does not reduce the valid TIP-20 balance range because total supply is already capped at `uint128::MAX`. The upper 128 bits are cache space attached to that balance slot, where `reserved` covers bits `136..255` and MUST be zero until assigned by a future TIP.

## Activation and Backward Compatibility

For pre-T6 execution, the slot MUST continue to be interpreted as the legacy raw balance. Cache fields MUST NOT be set pre-T6 activation.

At T6 activation, existing balance slots are interpreted as packed user-state words. This requires no migration because every valid TIP-20 balance fits in the low 128 bits. Existing holders therefore start with their previous balance and `rewardFlag == Uninitialized`.

For T6+ execution, TIP-20 APIs continue to expose balances normally. `balanceOf(account)` returns only the low 128-bit `balance` value, and no external interface exposes the packed cache fields directly.

## General User-State Cache Rules

> Cache fields are performance optimizations derived from existing protocol state. The protocol MUST ensure that cache is never stale, and always matches the source of truth.

Reviewer 3: Isn't a stale cache allowed, as long as it is set to uninitialized?

Author: technically yes, but the cache is supposed to be a state machine where the `Unitialized` state can't be reached again after the cache is set for the first time.

will update accordingly

Author: addressed in [`bb6d54e` (this PR)](https://github.com/tempoxyz/tempo/pull/4191/commits/bb6d54e3108a7cbebabb8c1679090539a5949eac)

Balance writes encode the new `balance` in the low 128 bits. Existing defined cache fields MAY be preserved, cleared, or updated according to their field-specific rules. Reserved bits MUST remain zero unless a future TIP assigns them. Future cache fields MAY use the reserved region only through a future TIP that defines their encoding, source of truth, invalidation behavior, and handling of invalid values. In practice this means that a packed state slot MAY have `balance == 0` while one of the cache fields is nonzero.

## Reward State Cache

`rewardFlag` is the first cache field defined in the user-state slot. It caches whether the account is opted into TIP-20 rewards so balance-changing paths can avoid unnecessary `SLOAD`s and mapping-key `keccak256` computations against reward-info storage. It can have the following values:

- `0`: `Uninitialized`.
- `1`: `OptedOut`.
- `2`: `OptedIn`.

> Any other encoded value is invalid and MUST be treated as `Uninitialized`.

Reviewer 2: ```suggestion
Any other encoded value is invalid and MUST be treated as a consensus error.
```

`rewardRecipient(account)` remains the source of truth for reward delegation and opt-in status, and the `rewardFlag` MUST always be aligned with it. As defined in TIP-20, `rewardRecipient != address(0)` means opted in, and `rewardRecipient == address(0)` means opted out.

The reward-state cache may be used by balance-changing paths that need to know whether a holder is opted into rewards, including transfers, mints, fee refunds, reward claims, and any internal balance movement that updates `optedInSupply`. These paths MAY use an initialized `rewardFlag` instead of reading reward info from separate mapping slots. When `rewardFlag` is `Uninitialized` or invalid, the implementation MUST read `rewardRecipient(account)` and derive the effective reward state from it.

> `setRewardRecipient(recipient)` MUST update `rewardRecipient`. It MUST also ensure that any existing `rewardFlag` cache for the account is not left stale: if the packed state slot is nonzero, `setRewardRecipient(recipient)` MUST update `rewardFlag` in that slot to match the new effective reward state before returning. If the packed state slot is zero, `setRewardRecipient(recipient)` MAY either leave it zero or create a cache-only packed state slot with `balance == 0` and `rewardFlag` matching the new effective reward state.

Reviewer 3: We can't really have "MAY" in the protocol implementation; this should probably be a MUST one way or the other? I'd say MUST that it updates the slot.

Author: addressed in [`bb6d54e` (this PR)](https://github.com/tempoxyz/tempo/pull/4191/commits/bb6d54e3108a7cbebabb8c1679090539a5949eac)

Reviewer 2: ```suggestion
`setRewardRecipient(recipient)` MUST update `rewardRecipient`. It MUST also ensure the `rewardFlag` cache for the account is not left stale: `setRewardRecipient(recipient)` MUST update `rewardFlag` in that slot to match the new effective reward state before returning.
```

authors: @0xrusowsky, Arsenii Kulikov (@klkvr)
status: Draft
related: TIP-20, TIP-1000
protocolVersion: T6
---

# TIP-1065: Repurpose TIP-20 Balance slots into User State

## Abstract

This TIP proposes reinterpreting each TIP-20 `balance` mapping value as a packed user-state word. The slot continues to store the user's token balance, but also gains space for small per-user cache fields. The initial cache field supports reward accounting, allowing balance-changing paths to avoid unnecessary storage reads and mapping-key (`keccak256`) computations.

## Motivation

The goal is to improve performance by minimizing `SLOAD`s and mapping-key `keccak256` computations in common TIP-20 execution paths.

TIP-20 already allocates one storage slot per holder with a nonzero balance. Reusing that slot as user state lets the protocol store small cache fields next to the balance without introducing new mappings or additional slots for active holders. Because TIP-20 total supply is capped at `uint128::MAX`, no account balance can exceed `uint128::MAX`; the upper 128 bits of each 256-bit balance slot are therefore unused by the balance itself.

> The first use case is TIP-20 reward accounting, which currently requires reading reward metadata from separate mapping slots during common balance-changing paths. Caching reward state in the existing balance slot reduces redundant storage reads and associated mapping hashing while keeping room in the same slot for future per-user caches.

Reviewer 4: not strictly necessary, but should we quantify somewhere in this tip that this saves x additional SLOADs on a transfer path after sender and receiver have been initialized?

Author: makes sense, will do

Author: added here [e89bb01](https://github.com/tempoxyz/tempo/pull/4191/commits/e89bb01f64323eee9de1e03aa8ca9a496cad48a6)

- `0`: `Uninitialized`.
- `1`: `OptedOut`.
- `2`: `OptedIn`.

Any other encoded value is invalid and MUST be treated as a consensus error.

`rewardRecipient(account)` remains the source of truth for reward delegation and opt-in status, and any initialized `rewardFlag` MUST always be aligned with it. As defined in TIP-20, `rewardRecipient != address(0)` means opted in, and `rewardRecipient == address(0)` means opted out.

The reward-state cache may be used by balance-changing paths that need to know whether a holder is opted into rewards, including transfers, mints, fee refunds, reward claims, and any internal balance movement that updates `optedInSupply`. These paths MAY use an initialized `rewardFlag` instead of reading reward info from separate mapping slots. When `rewardFlag` is `Uninitialized` or invalid, the implementation MUST read `rewardRecipient(account)` and derive the effective reward state from it.

Once `rewardFlag` has been initialized to `OptedOut` or `OptedIn`, it MUST NOT be reset to `Uninitialized`; future changes MUST move directly between initialized states.

> `setRewardRecipient(recipient)` MUST update `rewardRecipient`. It MUST also update the packed state slot before returning so `rewardFlag` matches the new effective reward state. If the packed state slot is zero, `setRewardRecipient(recipient)` MUST create a cache-only packed state slot with `balance == 0` and `rewardFlag` matching the new effective reward state.

Author: ```suggestion
`setRewardRecipient(recipient)` MUST update `rewardRecipient`. It MUST also ensure the `rewardFlag` cache for the account is not left stale: `setRewardRecipient(recipient)` MUST update `rewardFlag` in that slot to match the new effective reward state before returning.
```
